### DISCUSSION TITLE
Refactor front-end using Web Components

### DISCUSSION THREAD

**Discussion by zhuiks** [ezra-bible-app/ezra-bible-app#251] [Development]
Let's discuss here web components best practices and how they can be applied to Ezra Bible App

A few links that I found useful:
- Getting Started from [MDN](https://developer.mozilla.org/en-US/docs/Web/Web_Components)
- [All the Ways to Make a Web Component](https://webcomponents.dev/blog/all-the-ways-to-make-a-web-component/)
- Guidlines and utils: [Open WC](https://open-wc.org/docs/development/generator/) - I haven't tried it yet

**Comment by zhuiks** [comment]
## Whether to use ShadowDOM

**ShadowDOM**  acts like a DOM sub-tree and allows encapsulating styles, ids and classes. From my understanding there is not much difference on JavaScript side except adding extra `this.shadowRoot`. I'm not sure about the performance aspect. But it seems like a common practice is to use the following boilerplate:
``` javascript
const template = document.createElement('template');
template.innerHTML = `
  <style>
  </style>
  <div id="my-div"></div>
`;

class MyCounter extends HTMLElement {
  constructor() {
    super();
    this.attachShadow({ mode: 'open' });
  }

  connectedCallback() {
    this.shadowRoot.appendChild(template.content.cloneNode(true));
  }

}

customElements.define('my-element', MyElement);
```

**Comment by zhuiks** [comment]
## Should we use a Custom element's helper library

I think it doesn't matter for smaller components. But for something bigger it might be useful

Pros:
- event handlers in html
- auto binding between properties and attributes
- simplified component code and features such as @decorators
- up today best practices regarding custom elements (although some of those can lead into dead end as it happens with frameworks)
- Component libraries and design systems are already using one of the libraries, so there won't be additional increase in bundle size

Cons:
- additional learning curve
- increase in bundle size and possibly loading time
- It's not a language standard, so there is a risk that the library might die in the future


I'm considering these:
- [Lit Element](https://lit.dev/docs/getting-started/) - I think it's the most popular one. Came out from the Polymer elements
- [hybrids](https://hybrids.js.org/#/) - I like it's "functional" style

**Comment by tobias-klein** [comment]
Thanks for bringing up these points. I need to do some more of my own research before continuing this discussion.

**Comment by tobias-klein** [comment]
Here are some general thoughts ...

We need the following things:

1. A responsive base template / structure. This does not need to come right away, it could also be introduced later after having refactored a bit more at the "small component level". But eventually we need it. There are probably existing patterns/templates/structures available, we should not re-invent the wheel.
2. A discussion on whether we use web components for "everything" or for example just at the lowest level of the element hierarchy ("widgets" / smaller components). I am a bit hesitant right now to use web components for "composing containers" (containers where you just combine a bunch of other web components).
3. A discussion on whether we stick with Vanilla JS or whether we introduce additional helper libraries. The web has been extremely dynamic in the last 20 years. Helper libraries may die pretty quickly ... whatever we choose today must be sustainable at least for the next 5-8 years I would say.
4. A discussion about the use cases that we have for "data binding" (elements updating automatically somewhere once the source value has changed).
5. When considering helper libraries we should not only look at the frontend aspect, but also at things like data stores. Something that has peaked my interest is [Redux](https://redux.js.org). Currently the "data stores" related to the tagging function are completely self-made, but I am not entirely happy with it.
6. Performance considerations. How is performance impacted if you have a lot of web components in the Dom? (Would be the case if you use a separate web component for each verse box or tag element)

**Reply by zhuiks** [reply]
> When considering helper libraries we should not only look at the frontend aspect, but also at things like data stores. Something that has peaked my interest is Redux. Currently the "data stores" related to the tagging function are completely self-made, but I am not entirely happy with it.

Unless we use a big framework ([Svetle](https://svelte.dev/) is the newer one with native component-style approach) discussion about state and data stores worth a separate thread

**Comment by zhuiks** [comment]
### Responsive base template
> A responsive base template / structure. This does not need to come right away, it could also be introduced later after having refactored a bit more at the "small component level". But eventually we need it. There are probably existing patterns/templates/structures available, we should not re-invent the wheel.

It's just a simple grid/flex layout implemented with css and html. It can be a content of the main `index.html`. Maybe something like this:
```html
<header>
  <settings-component></settings-component>
</header>
<main>
  <aside>
    <tags-component></tags-component>
    <dictionary-component></dictionary-component>
  </aside>
  <tabs-component></tabs-component>
  <footer>
    <interlinear-component></interlinear-component>
  </footer>
</main>
```

The main idea to have all the components decoupled. So we can move for example `<dictionary-component>` deeper inside `<tabs-component>` (to be tab specific) without any need of refactoring

**Reply by zhuiks** [reply]
The continued discussion about "responsive base  template" is happening under #299 and being implemented with PR #302

**Comment by zhuiks** [comment]
### Use web components for "everything"
> A discussion on whether we use web components for "everything" or for example just at the lowest level of the element hierarchy ("widgets" / smaller components). I am a bit hesitant right now to use web components for "composing containers" (containers where you just combine a bunch of other web components).

The main benefit of the web component is encapsulating and decoupling logic. I think even if it's just encapsulating styles it is still worth it. The result will be in cleaner and more clear code.

> Performance considerations. How is performance impacted if you have a lot of web components in the Dom? (Would be the case if you use a separate web component for each verse box or tag element)

I think it all depends on the implementation of a component and what's happening inside `constructor()` and `connectedCallback()`.
Practically the question is about how to properly implement `verse-box` keeping in mind that there could be more than 1000 of those.

**Comment by tobias-klein** [comment]
Note that as part of the dynamic book loading feature I have implemented a new web component that inherits from the standard select element, but formats it using jQuery UI selectmenu + hooks it up with the option persistance system.

See https://github.com/ezra-bible-app/ezra-bible-app/blob/feature-chapter-loading/app/frontend/components/select_option.js

**Comment by zhuiks** [comment]
I'm trying to refactor `installModuleAssistant` with web components (PR #317)

I'll do it from my understanding of utilizing `<template>` and `shadowDOM`. Later we can discuss it and select the best approach.
In the end of the day, I think, having consistency in code is the most important
